source/s3_endpoint.c

```diff
@@ -278,11 +317,13 @@ static void s_s3_endpoint_release(struct aws_s3_endpoint *endpoint) {
     /* END CRITICAL SECTION */
 
     if (should_destroy) {
-        /* The endpoint may have async cleanup to do (connection manager).
-         * When that's all done we'll invoke a completion callback.
-         * Since it's a crime to hold a lock while invoking a callback,
-         * we make sure that we've released the client's lock before proceeding... */
-        s_s3_endpoint_ref_count_zero(endpoint);
+        if (client_active) {
+            /* schedule the cleanup task */
+            aws_s3_client_schedule_process_work(endpoint->client);
+        } else {
+            /* do a sync cleanup since client is not active anymore to avoid any cleanup delay */
+            s_s3_endpoint_ref_count_zero(endpoint);
+        }
```

Comment:
Question/Confusion: Is it worth avoiding `aws_event_loop_schedule_task_now` here? If we simplify the code here to always trigger `aws_s3_client_schedule_process_work`, we will clean up the endpoint on the client process work thread, but does it matter?